src/Sensors/NotificationSensor.php

```diff
@@ -2,38 +2,57 @@
 
 namespace Laravel\Nightwatch\Sensors;
 
+use Illuminate\Notifications\Events\NotificationSending;
 use Illuminate\Notifications\Events\NotificationSent;
 use Laravel\Nightwatch\Clock;
 use Laravel\Nightwatch\Records\Notification;
 use Laravel\Nightwatch\State\CommandState;
 use Laravel\Nightwatch\State\RequestState;
 use Laravel\Nightwatch\Types\Str;
+use RuntimeException;
 
 use function hash;
+use function round;
 use function str_contains;
 
 /**
  * @internal
  */
 final class NotificationSensor
 {
+    private ?float $startTime = null;
+
+    private ?int $duration = null;
+
     public function __construct(
         private RequestState|CommandState $executionState,
         private Clock $clock,
     ) {
         //
     }
 
-    public function __invoke(NotificationSent $event): void
+    public function __invoke(NotificationSending|NotificationSent $event): void
     {
         $now = $this->clock->microtime();
 
+        if ($event instanceof NotificationSending) {
+            $this->startTime = $now;
+            $this->duration = null;
+
+            return;
+        }
+
+        if ($this->startTime === null) {
+            throw new RuntimeException('No start time found for ['.$event->notifiable::class.'].'); // @phpstan-ignore classConstant.nonObject
+        }
+
         if (str_contains($event->notification::class, "@anonymous\0")) {
             $class = Str::before($event->notification::class, "\0");
         } else {
             $class = $event->notification::class;
         }
 
+        $this->duration ??= (int) round(($now - $this->startTime) * 1_000_000);
```

Comment:
Do we need the `??=` operator here?

Looking at this, it seems we would always want to recompute the duration at this point for notifications?

Reply:
No, we don't. I can't imagine a scenario where we want to reuse the same duration. I've updated the code. Thank you for spotting this!